src/unix/filesystem.c

```diff
@@ -0,0 +1,150 @@
+#include <unix_internal.h>
+#include <filesystem.h>
+
+#define SYMLINK_HOPS_MAX    8
+
+// fused buffer wrap, split, and resolve
+int resolve_cstring(tuple cwd, const char *f, tuple *entry, tuple *parent)
+{
+    if (!f)
+        return -EFAULT;
+
+    tuple t = *f == '/' ? filesystem_getroot(current->p->fs) : cwd;
+    tuple p = t;
+    buffer a = little_stack_buffer(NAME_MAX);
+    char y;
+    int nbytes;
+    int err;
+
+    while ((y = *f)) {
+        if (y == '/') {
+            if (buffer_length(a)) {
+                p = t;
+                t = lookup(t, intern(a));
+                if (!t)
+                {
+                    err = -ENOENT;
+                    goto done;
+                }
+                err = filesystem_follow_links(t, p, &t);
+                if (err) {
+                    t = false;
+                    goto done;
+                }
+                if (!children(t))
+                    return -ENOTDIR;
+                buffer_clear(a);
+            }
+            f++;
+        } else {
+            nbytes = push_utf8_character(a, f);
+            if (!nbytes) {
+                thread_log(current, "Invalid UTF-8 sequence.\n");
+                return -ENOENT;
+            }
+            f += nbytes;
+        }
+    }
+
+    if (buffer_length(a)) {
+        p = t;
+        t = lookup(t, intern(a));
+    }
+    err = -ENOENT;
+done:
+    if (!t && (*f == '/') && (*(f + 1)))
+        /* The path being resolved contains entries under a non-existent
+         * directory. */
+        p = false;
+    if (parent)
+        *parent = p;
+    if (entry)
+        *entry = t;
+    return (t ? 0 : err);
+}
+
+int resolve_cstring_follow(tuple cwd, const char *f, tuple *entry,
+        tuple *parent)
+{
+    tuple t, p;
+    int ret = resolve_cstring(cwd, f, &t, &p);
+    if (!ret) {
+        ret = filesystem_follow_links(t, p, &t);
+    }
+    if (entry) {
```

Comment:
Indeed setting *entry doesn't make much sense, because if ret != 0 it means the target tuple hasn't been found. But setting *parent is necessary for those cases (e.g. file and dir creation) where we need to know the parent of a filesystem entry that doesn't exist.